analyzers/src/SonarAnalyzer.Common/Rules/HardcodedIpAddressBase.cs

```diff
@@ -30,64 +30,86 @@
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected HardcodedIpAddressBase(IAnalyzerConfiguration analyzerConfiguration, System.Resources.ResourceManager rspecResources) : base(analyzerConfiguration)
         {
+            rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, rspecResources).WithNotConfigurable();
         }
 
-        protected Action<SyntaxNodeAnalysisContext> GetAnalysisAction(DiagnosticDescriptor rule) =>
-            c =>
+        protected override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterSyntaxNodeActionInNonGenerated(GeneratedCodeRecognizer, CheckForHardcodedIpAddresses, SyntaxKind);
+
+        private void CheckForHardcodedIpAddresses(SyntaxNodeAnalysisContext context)
+        {
+            if (!IsEnabled(context.Options))
             {
-                var stringLiteral = (TLiteralExpression)c.Node;
-                var literalValue = GetValueText(stringLiteral);
-                if (IsLocalHost(literalValue))
-                {
-                    return;
-                }
-
-                var variableName = GetAssignedVariableName(stringLiteral);
-                if (variableName != null &&
-                    IgnoredVariableNames.Any(variableName.Contains))
-                {
-                    return;
-                }
-
-                if (HasAttributes(stringLiteral))
-                {
-                    return;
-                }
-
-                if (!IPAddress.TryParse(literalValue, out var address))
-                {
-                    return;
-                }
-
-                if (address.AddressFamily == AddressFamily.InterNetwork &&
-                    literalValue.Split('.').Length != 4)
-                {
-                    return;
-                }
-
-                c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, stringLiteral.GetLocation(), literalValue));
-            };
+                return;
+            }
+
+            var stringLiteral = (TLiteralExpression)context.Node;
+            var variableName = GetAssignedVariableName(stringLiteral);
+
+            if (variableName != null && ignoredVariableNames.Any(x => variableName.IndexOf(x, StringComparison.InvariantCultureIgnoreCase) >= 0))
+            {
+                return;
+            }
 
-        private static bool IsLocalHost(string s) =>
-            s == "::" || s == "127.0.0.1";
+            if (HasAttributes(stringLiteral))
+            {
+                return;
+            }
+
+            var literalValue = GetValueText(stringLiteral);
+
+            IPAddress address;
+
+            if (!IPAddress.TryParse(literalValue, out address))
+            {
+                return;
+            }
+
+            if (IPAddress.IsLoopback(address)
+                || address.GetAddressBytes().All(x => x == 0)
```

Comment:
This line deserves a short explanation now, as it's not that clear without the original variable names
```suggestion
                || address.GetAddressBytes().All(x => x == 0) // Nonroutable 0.0.0.0 or 0::0
```